[TypeProf][PGO]Support skipping vtable comparisons for a class and its derived ones

mingmingl-llvm opened the pull request:
Performance critical core libraries could be highly-optimized for arch or micro-arch features. For instance, the absl crc library specializes different templated classes among different hardwares [1]. In a practical setting, it's likely that instrumented profiles are collected on one type of machine and used to optimize binaries that run on multiple types of hardwares.

While this kind of specialization is rare in terms of lines of code, compiler can do a better job to skip vtable-based ICP.
* The per-class `Extend` implementation is arch-specific as well. If an instrumented profile is collected on one arch and applied to another arch where `Extend` implementation is different, `Extend` might be regarded as unlikely function in the latter case. `ABSL_ATTRIBUTE_HOT` annotation alleviates the problem by putting all `Extend` implementation into the hot text section [2]

This change introduces a comma-separated list to specify the mangled vtable names, and ICP pass will skip vtable-based comparison if a vtable variable definition is shown to be in its class hierarchy (per LLVM type metadata).

[1] https://github.com/abseil/abseil-cpp/blob/c6b27359c3d27438b1313dddd7598914c1274a50/absl/crc/internal/crc_x86_arm_combined.cc#L621-L650
[2] https://github.com/abseil/abseil-cpp/blame/c6b27359c3d27438b1313dddd7598914c1274a50/absl/crc/internal/crc_x86_arm_combined.cc#L370C3-L370C21

mingmingl-llvm committed:
[TypeProf][PGO]Support specification of vtables for which type profiles are known to be unrepresentative

snehasish commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -132,6 +132,11 @@ static cl::opt<int> ICPMaxNumVTableLastCandidate(
     cl::desc("The maximum number of vtable for the last candidate."));
 
+static cl::opt<std::string> ICPKnownUnrepresentativeVTables(
+    "icp-known-unrepresentative-vtables", cl::init(""), cl::Hidden,
```
Do we need to make this option and the code (e.g. variable names, debug messages) specific to "profile unrepresentativeness"? Can we instead generalize it like naming the option "icp-ignore-vtables", the variable can be `DenseSet<StringRef> IgnoredBaseTypes` and so on?

I think the motivation is justified and can be mentioned in the comments and cl::desc but the implementation don't need to be. IMO this makes the code a little easier to read and follow. What do you think?

david-xl commented:
I agree that the option name can be more general as there can be other use cases too.

mingmingl-llvm committed:
resolve review feedback

snehasish commented on llvm/lib/Support/CommandLine.cpp:
```diff
@@ -2060,6 +2062,24 @@ bool parser<float>::parse(Option &O, StringRef ArgName, StringRef Arg,
+    OptionValue<DenseSet<StringRef>> D, size_t GlobalWidth) const {}
+
+bool parser<DenseSet<StringRef>>::parse(Option &O, StringRef ArgName,
+                                        StringRef Arg,
```
Are you certain that the contents of `StringRef Arg` outlives the DenseSet<StringRef> returned? All the others above return the contents by value.

I would prefer a cl::opt<std::string> which is parsed and validated in the ICP pass. Is there an advantage to your proposed approach and are there any precedents in the codebase? IMO this option feels overly restrictive when rejecting duplicated strings from the user.

snehasish commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -851,8 +859,27 @@ bool IndirectCallPromoter::isProfitableToCompareVTables(
+        if (auto *TypeId = dyn_cast<MDString>(Type->getOperand(1).get()))
+          if (VTableSet.contains(TypeId->getString().str())) {
+            LLVM_DEBUG(dbgs()
+                       << "    vtable profiles are known to be "
```
Update the comment? s/known to be unrepresentative/ignored by option ..../

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -132,6 +132,14 @@ static cl::opt<int> ICPMaxNumVTableLastCandidate(
     "icp-max-num-vtable-last-candidate", cl::init(1), cl::Hidden,
     cl::desc("The maximum number of vtable for the last candidate."));
 
+static cl::opt<DenseSet<StringRef>> ICPIgnoredBaseTypes(
```
cl::list is the typical way of supporting an option that takes a list. Can you just use that instead of adding another cl::opt parser?

snehasish commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -851,8 +859,27 @@ bool IndirectCallPromoter::isProfitableToCompareVTables(
 
     uint64_t CandidateVTableCount = 0;
-    for (auto &[GUID, Count] : VTableGUIDAndCounts)
+    SmallVector<MDNode *, 2> Types;
```
We can move this to line L870 and drop the `clear()`. Since it's a small vector with only 2 pointers (on the stack) I don't expect changing this will affect performance.

teresajohnson reviewed (commented):
Is the issue here specific to vtable based ICP or does it also arise in the function based ICP?

mingmingl-llvm committed:
1. use cl::list<std::string> for option type
2. Update `-icp-ignored-base-types` with one base type since this is
   sufficient

minglotus-6 commented:
> Is the issue here specific to vtable based ICP or does it also arise in the function based ICP?

This also arises to function-based ICP, and the `Extend` method of class template is marked as [hot](https://github.com/abseil/abseil-cpp/blame/c6b27359c3d27438b1313dddd7598914c1274a50/absl/crc/internal/crc_x86_arm_combined.cc#L370C3-L370C21) with `ABSL_ATTRIBUTE_HOT` to at least put `Extend` in `.text.hot` section (rather than `.text.unlikely`)

snehasish commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -132,6 +132,15 @@ static cl::opt<int> ICPMaxNumVTableLastCandidate(
+        "and their derived ones will not be vtable-ICP'ed. Useful when the "
+        "profiled types and actual types in the optimized binary could be "
+        "different due to profiling "
+        "limitations."));
```
nit: could be on previous line

minglotus-6 replied:
done.

snehasish commented on llvm/test/Transforms/PGOProfile/icp_vtable_cmp.ll:
```diff
@@ -2,6 +2,7 @@
 
 ; RUN: opt < %s -passes='pgo-icall-prom' -pass-remarks=pgo-icall-prom -enable-vtable-profile-use -icp-max-num-vtable-last-candidate=2 -S 2>&1 | FileCheck %s --check-prefixes=VTABLE-COMMON,VTABLE-CMP
 ; RUN: opt < %s -passes='pgo-icall-prom' -pass-remarks=pgo-icall-prom -enable-vtable-profile-use -icp-max-num-vtable-last-candidate=1 -S 2>&1 | FileCheck %s --check-prefixes=VTABLE-COMMON,FUNC-CMP
+; RUN: opt < %s -passes='pgo-icall-prom' -pass-remarks=pgo-icall-prom -enable-vtable-profile-use -icp-max-num-vtable-last-candidate=1 -icp-ignored-base-types='Base1' -S 2>&1 | FileCheck %s --check-prefixes=VTABLE-COMMON,FUNC-CMP
```
Add a comment to clarify the expectations for this test?

minglotus-6 replied:
done.

mingmingl-llvm committed:
resolve review feedback

snehasish reviewed (approved):
lgtm

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -851,8 +863,25 @@ bool IndirectCallPromoter::isProfitableToCompareVTables(
+        if (auto *TypeId = dyn_cast<MDString>(Type->getOperand(1).get()))
+          if (IgnoredBaseTypes.contains(TypeId->getString())) {
+            LLVM_DEBUG(dbgs() << "    vtable profiles should be ignored. Bail "
+                                 "out vtable comparison.");
```
s/out/out of/

minglotus-6 replied:
done.

teresajohnson commented on llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp:
```diff
@@ -851,8 +863,25 @@ bool IndirectCallPromoter::isProfitableToCompareVTables(
+
+    for (auto &[GUID, Count] : VTableGUIDAndCounts) {
       CandidateVTableCount += Count;
+      auto *VTableVar = Symtab->getGlobalVariable(GUID);
```
can we skip all the new handling when IgnoredBaseTypes is empty?

minglotus-6 replied:
This makes sense.

I took the liberty to use a helper function for the new handling, to simplify the control flow (e.g., continue, inner-for-loop and return) inside the outer loop.

mingmingl-llvm committed:
1. resolve review feedback 2. Update option description to mention the
   strings are type info names (the string literals in llvm type
   metadata)
